include/xtensor/xcontainer.hpp

```diff
@@ -374,17 +380,12 @@ namespace xt
         adapt_strides();
     }
 
-    /**
-     * Transposes the container inplace by permuting the shape with @permutation.
-     * @param permutation the dimensions with the given permutation
-     */
     template <class D>
-    inline void xcontainer<D>::transpose(const std::vector<size_type>& permutation)
+    inline void xcontainer<D>::transpose(const std::vector<size_type>& permutation, const do_check check_policy)
     {
-        if (permutation.size() != m_shape.size())
-        {
-            throw transpose_error("Permutation does not have the same size as shape");
-        }
+        // casting to void eliminates unused parameter warning plus optimizes compiler output
+        // according to https://arobenko.gitbooks.io/bare_metal_cpp/content/compiler_output/tag_dispatch.html
+        static_cast<void>(check_policy);
```

Comment:
This trick is overkilled here. Removing a assembly instruction is useful when the rest of your function is translated in a few lines of assembly and called very often (like an access operator in a loop for instance). But that's not the case here (even if the private function, let's call it `transpose_impl`, is not inlined at the call site), you have loops, throw instruction, etc, in the private function. Thus you can improve the code readability without any performance hit: do not name the `Tag` parameter in the function, no `const` classifier, and no `static_cast`:
```cpp
inline void xcontainer<D>::transpose_impl(const std::vector<size_type>& permutation, transpose_policy::full_check)
{
    // no static_cast<void> here
    // do the checks here
    transpose_impl(permutation, transpose_policy::no_check());
}
```